### PR TITLE
Selective dequantization

### PR DESCRIPTION
**Pull request by RezaYazdaniAminabadi**
This PR adds a new functionality for the dequantizer function, called `selective_dequantize`, which enables partially dequantizing a 3-dimensional matrix in case we don't need to dequantize all the data from lower bit (like fp8/fp6) to bf16.
I also added a unit test to check its functionality.

### CODE REVIEW

**Comment by RezaYazdaniAminabadi** [comment]
> @RezaYazdaniAminabadi please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement


@microsoft-github-policy-service agree company="Snowflake

**Comment by RezaYazdaniAminabadi** [comment]
@JamesTheZ could i please get your review on this PR too? thanks

**Comment by sfc-gh-reyazda** [comment]
Hi @loadams can we please have someone review this so we merge it soon? thanks :)

File: csrc/fp_quantizer/quantize.cpp [hunk 1]
```diff
@@ -78,8 +78,39 @@ void dequantize(torch::Tensor& val,
...
+        return;                                                                             \
+    }
+void selective_dequantize(torch::Tensor& val,
+                          torch::Tensor& val_q,
+                          torch::Tensor& indexes,
+                          int group_size,
+                          int q_mantisa_bits,
+                          int q_exponent_bits)
+{
+    int total_elems = at::numel(val);
```
Comment (JamesTheZ) [inline csrc/fp_quantizer/quantize.cpp:101]: Why count the elements of the original tensor rather than the sliced tensor?

Comment (RezaYazdaniAminabadi) [reply csrc/fp_quantizer/quantize.cpp:101]: the first argument for the `selective_dequantize` is actually the tensor in which we dequantize the selected indexes. The second one `val_q` is the quantized data.

File: csrc/fp_quantizer/quantize.cu [hunk 2]
```diff
@@ -393,3 +394,137 @@ void launch_dequantization(uint8_t* val,
...
+    const uint32_t group_size_bytes = (group_size * quantized_bits / 8);
+    const uint8_t* load_base_ptr =
+        val + g_index * (group_size_bytes + 4) + (input_index % group_size) * quantized_bits / 8;
+
+    int mantisa_mask = ((1 << q_mantisa_bits) - 1);
+    mantisa_mask <<= (_mantisa_bits - q_mantisa_bits);
+
+    T* store_base_ptr = q_val + tidx + blockIdx.x * total_num_elements;
+    float scale;
+
```
[the remark this answers is not in the document]
Comment (RezaYazdaniAminabadi) [reply csrc/fp_quantizer/quantize.cu:428]: right, I will clean this part in another PR, thanks @JamesTheZ

File: csrc/fp_quantizer/quantize.cpp (hunk 1 above)
Comment (JamesTheZ) [reply csrc/fp_quantizer/quantize.cpp:101]: Ah I see. It is processed in the python API. Thanks.

**Review by JamesTheZ — APPROVED** [review]

**Comment by sfc-gh-reyazda** [comment]
> @microsoft-github-policy-service agree [company="{your company}"]

@microsoft-github-policy-service agree [company="Snowflake"]

**Pull request merged**